### PR TITLE
Add NGINX reverse proxy

### PR DESCRIPTION
**Pull request by surminus**
I tried in vain to use Traefik as a small reverse proxy since it forwarded error pages out of the box, and is built for cloud environments.

Unfortunately I ran into issues with the `Authorization` header being set when requesting the error page, and S3 failing since that header was set.

This replaces the functionality with Good Ole NGINX, and strips the `Authorization` header when we're passing on a request to an error page.

I've defined every single error page in a separate block with a separate `upstream` block for the static pages, since I ran into issues trying to be clever, where I had to set `resolver` configuration.

Basically, I got bored and decided not to be clever.

### CODE REVIEW

**Comment by surminus** [comment]
One thing I've realised is we may need to set real ip configuration? Lots of stuff I was hoping to remove has sadly come back to haunt us with the introduction of NGINX.

**Comment by elenatanasoiu** [comment]
Followed by:
- https://github.com/futurelearn/deployment/pull/81
- https://github.com/futurelearn/terraform/pull/257

Related to:
- https://github.com/futurelearn/architecture-decision-record/pull/4

File: nginx-proxy/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
+FROM nginx
+
+RUN apt-get update && apt-get install -y netcat
```
Comment (elenatanasoiu) [inline nginx-proxy/Dockerfile]: What are we doing with `netcat`?

File: nginx-proxy/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+COPY entrypoint.sh /opt/entrypoint.sh
+RUN chmod +x /opt/entrypoint.sh
+
+COPY default.conf /etc/nginx/default.conf.template
+
+ENTRYPOINT ["/opt/entrypoint.sh"]
+
+CMD ["nginx", "-g", "daemon off;"]
```
Comment (elenatanasoiu) [inline nginx-proxy/Dockerfile:12]: Will this be changed to `daemon on` at some point? Or is the idea to have this container just running `nginx` so it should be in the foreground all the time?

File: nginx-proxy/README.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# nginx-proxy
+
+A small reverse to forward error pages to an external location.
```
Comment (elenatanasoiu) [inline nginx-proxy/README.md]: reverse-proxy* ?

File: nginx-proxy/default.conf [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+  access_log /dev/stdout log_json;
+
+  proxy_intercept_errors on;
+  proxy_ssl_server_name on;
+
+  proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+  proxy_set_header X-Real-IP  $remote_addr;
+  proxy_set_header Host $http_host;
+  proxy_set_header If-Modified-Since $http_if_modified_since;
```
Comment (elenatanasoiu) [inline nginx-proxy/default.conf:46]: I understand we need to overwrite the `Authorization` header and `Host` header for each error page. Why are we setting these particular 4 headers as well?

File: nginx-proxy/entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+if [[ -z "${STATIC_PAGES_HOST}" ]]; then
+  echo "ERROR: Must set STATIC_PAGES_HOST environment variable"
+  exit 1
+fi
+
+[[ -z "${PORT}" ]]         && export PORT="80"
+[[ -z "${BACKEND_HOST}" ]] && export BACKEND_HOST="127.0.0.1"
+
+# shellcheck disable=SC2016
+envsubst '${BACKEND_HOST} ${STATIC_PAGES_HOST} ${PORT}' < /etc/nginx/default.conf.template > /etc/nginx/conf.d/default.conf
```
Comment (elenatanasoiu) [inline nginx-proxy/entrypoint.sh:13]: Nice! TIL

File: nginx-proxy/Dockerfile (hunk 1 above)
Comment (elenatanasoiu) [reply nginx-proxy/Dockerfile]: Ah https://github.com/futurelearn/deployment/pull/81/commits/f5a19f350b0f541dd0b92387980d29d0b66ff9a3#diff-fcc9f67616f16fa4f6a6196995e4c3ccR100

Comment (surminus) [reply nginx-proxy/Dockerfile]: Yeah, healthchecks.

File: nginx-proxy/Dockerfile (hunk 2 above)
Comment (surminus) [reply nginx-proxy/Dockerfile:12]: It should run as a foreground process in a container. The container should basically run a single process, so there is no need for it to run in the background.

File: nginx-proxy/default.conf (hunk 4 above)
Comment (surminus) [reply nginx-proxy/default.conf:46]: I copied this from our [current NGINX config](https://github.com/futurelearn/chef/blob/master/site-cookbooks/app/templates/default/futurelearn_nginx_conf.erb#L141-L144).

The main thing is `X-Forwarded-For` - I wanted to ensure we continued forwarding the remote user IP.

**Review by elenatanasoiu — APPROVED** [review]

**Comment by surminus** [comment]
I added a commit for real ip configuration

**Review by elenatanasoiu — APPROVED** [review]
